libvast/src/system/connector.cpp

```diff
@@ -0,0 +1,256 @@
+  if (!deadline)
+    return caf::infinite;
+  const auto now = std::chrono::steady_clock::now();
+  if (now >= *deadline)
+    return std::nullopt;
+  return *deadline - now;
+}
+
+bool should_retry(const caf::error& err,
+                  const std::optional<caf::timespan>& remaining_time,
+                  caf::timespan delay) {
+  return remaining_time && *remaining_time > delay && is_recoverable_error(err);
+}
+
+std::string format_time(caf::timespan timespan) {
+  if (caf::is_infinite(timespan))
+    return "infinite";
+  std::vector<std::string> measurements;
+  if (auto h = std::chrono::duration_cast<std::chrono::hours>(timespan);
+      h.count() > 0) {
+    measurements.emplace_back(fmt::format("{}", h));
+    timespan -= std::chrono::duration_cast<caf::timespan>(h);
+  }
+  if (auto m = std::chrono::duration_cast<std::chrono::minutes>(timespan);
+      m.count() > 0) {
+    measurements.emplace_back(fmt::format("{}", m));
+    timespan -= std::chrono::duration_cast<caf::timespan>(m);
+  }
+  if (auto s = std::chrono::duration_cast<std::chrono::seconds>(timespan);
+      s.count() > 0) {
+    measurements.emplace_back(fmt::format("{}", s));
+    timespan -= std::chrono::duration_cast<caf::timespan>(s);
+  }
+  if (auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(timespan);
+      ms.count() > 0) {
+    measurements.emplace_back(fmt::format("{}", ms));
+    timespan -= std::chrono::duration_cast<caf::timespan>(ms);
+  }
+
+  return fmt::format("{}", fmt::join(measurements, " "));
+}
+
+void log_connection_failed(caf::timespan remaining_time,
+                           caf::timespan retry_delay) {
+  VAST_INFO("remote node connection failed. Retrying to "
+            "connect in {} (remaining time: {})",
+            format_time(retry_delay), format_time(remaining_time));
+}
+
+connector_actor::behavior_type make_no_retry_behavior(
+  connector_actor::stateful_pointer<connector_state> self,
+  std::optional<std::chrono::steady_clock::time_point> deadline) {
+  return {
+    [self, deadline](atom::connect,
+                     connect_request request) -> caf::result<node_actor> {
+      const auto remaining_time = calculate_remaining_time(deadline);
+      if (!remaining_time)
+        return caf::make_error(ec::timeout,
+                               fmt::format("{} couldn't connect to VAST node"
+                                           "within a given deadline",
+                                           *self));
+      auto rp = self->make_response_promise<node_actor>();
+      self
+        ->request(self->state.middleman, *remaining_time, caf::connect_atom_v,
+                  request.host, request.port)
+        .then(
+          [rp, request](const caf::node_id& node_id,
+                        caf::strong_actor_ptr& node,
+                        const std::set<std::string>&) mutable {
+            VAST_INFO("client connected to VAST node: {} at {}:{}",
```

Comment:
```suggestion
            VAST_INFO("client connected to VAST node at {}:{}",
```

I think the `node_id` is too much of an internal to be shown in a log message on INFO level. Maybe you can add a separate DEBUG message that shows it.